pennylane/gradients/metric_tensor.py

```diff
@@ -47,25 +51,19 @@ def _contract_metric_tensor_with_cjac(mt, cjac, tape):  # pylint: disable=unused
     if isinstance(mt, tuple) and len(mt) == 1:
         mt = mt[0]
     if isinstance(cjac, tuple):
-        # Classical processing of multiple arguments is present. Return cjac.T @ mt @ cjac
+        # Classical processing of multiple arguments is present. Return cjac[i].T @ mt @ cjac[i]
         # as a tuple of contractions.
-        metric_tensors = tuple(
-            qml.math.tensordot(c, qml.math.tensordot(mt, c, axes=[[-1], [0]]), axes=[[0], [0]])
-            for c in cjac
-            if c is not None
-        )
+        metric_tensors = tuple(_mt_cjac_tdot(mt, c) for c in cjac if c is not None)
         return metric_tensors[0] if len(metric_tensors) == 1 else metric_tensors
 
-    is_square = cjac.shape == (1,) or (cjac.ndim == 2 and cjac.shape[0] == cjac.shape[1])
+    if not qml.math.is_abstract(cjac):
+        is_square = cjac.shape == (1,) or (cjac.ndim == 2 and cjac.shape[0] == cjac.shape[1])
 
-    if is_square and qml.math.allclose(cjac, qml.numpy.eye(cjac.shape[0])):
-        # Classical Jacobian is the identity. No classical processing
-        # is present inside the QNode.
-        return mt
-    mt_cjac = qml.math.tensordot(mt, cjac, axes=[[-1], [0]])
-    mt = qml.math.tensordot(cjac, mt_cjac, axes=[[0], [0]])
+        if is_square and qml.math.allclose(cjac, qml.numpy.eye(cjac.shape[0])):

```

Comment:
We should probably not use `qml.numpy` here just for the sake of good practice 😅😅
```suggestion
        if is_square and qml.math.allclose(cjac, qml.math.eye(cjac.shape[0])):
```